engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala

```diff
@@ -0,0 +1,157 @@
+import org.enso.interpreter.runtime.Module
+
+/**
+  * A utility structure for resolving symbols in a given module.
+  * @param types the types defined in the current module
+  * @param currentModule the module holding these bindings
+  */
+case class BindingsMap(
+  types: List[BindingsMap.Cons],
+  currentModule: Module
+) extends IRPass.Metadata {
+  import BindingsMap._
+
+  override val metadataName: String = "Bindings Map"
+
+  override def duplicate(): Option[IRPass.Metadata] = Some(this)
+
+  /**
+    * Other modules, imported by [[currentModule]].
+    */
+  var resolvedImports: List[Module] = List()
+
+  private def findConstructorCandidates(
+    name: String
+  ): List[ResolvedConstructor] = {
+    types
+      .filter(_.name == name)
+      .map(ResolvedConstructor(currentModule, _))
+  }
+
+  private def findLocalCandidates(name: String): List[ResolvedName] = {
+    if (currentModule.getName.item == name) {
+      List(ResolvedModule(currentModule))
+    } else {
+      findConstructorCandidates(name)
+    }
+  }
+
+  private def findQualifiedImportCandidates(
+    name: String
+  ): List[ResolvedName] = {
+    resolvedImports.filter(_.getName.item == name).map(ResolvedModule)
+  }
+
+  private def findExportedCandidatesInImports(
+    name: String
+  ): List[ResolvedName] = {
+    resolvedImports
+      .map { mod =>
+        mod.getIr.unsafeGetMetadata(
+          BindingResolution,
+          "Wrong pass ordering. Running resolution on an unparsed module"
+        )
+      }
+      .flatMap(_.findConstructorCandidates(name))
+  }
+
+  private def handleAmbiguity(
+    candidates: List[ResolvedName]
+  ): Either[ResolutionError, ResolvedName] = {
+    candidates match {
+      case List()   => Left(ResolutionNotFound)
+      case List(it) => Right(it)
+      case items    => Left(ResolutionAmbiguous(items))
+    }
+  }
+
+  /**
+    * Resolves a name in the context of current module.
+    * @param name the name to resolve.
+    * @return a resolution for [[name]] or an error, if the name could not be
+    *         resolved.
+    */
+  def resolveUppercaseName(
+    name: String
+  ): Either[ResolutionError, ResolvedName] = {
+    val local = findLocalCandidates(name)
+    if (local.nonEmpty) {
+      return handleAmbiguity(local)
+    }
+    val qualifiedImps = findQualifiedImportCandidates(name)
+    if (qualifiedImps.nonEmpty) {
+      return handleAmbiguity(qualifiedImps)
+    }
+    handleAmbiguity(findExportedCandidatesInImports(name))
+  }
+}
+
+object BindingsMap {
+
+  /**
+    * A representation of a constructor.
+    *
+    * @param name the name of the constructor.
+    * @param arity the number of fields in the constructor.
+    */
+  case class Cons(name: String, arity: Int)
+
+  /**
+    * A result of successful name resolution.
+    */
+  sealed trait ResolvedName
+
+  /**
+    * A representation of a name being resolved to a constructor.
+    * @param definitionModule the module the constructor is defined in.
```

Comment:
```suggestion
    * A representation of a name being resolved to a constructor.
    *
    * @param definitionModule the module the constructor is defined in.
```
Please check your docs for the missing newlines.